lib/pbench/server/api/resources/query_apis/index_search.py

```diff
@@ -0,0 +1,230 @@
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(self, es_json: JSON) -> JSON:
+        """
+        Helper function for Postprocess for flattening the json. Any exception raised here will be
+        treated as a postprocess error.
+        We convert a nested data layer object into a new object with only one layer of key/value pairs.
+        Objects such as parent -> child are flattened as parent.child like following:
+
+        parent -> array -> [ele1, ele2] are flattened as
+                parent.array.0 = ele1, parent.array.1 = ele2
+
+        parent -> array -> [{key1: value1}, {key2: value2}] are flattened as
+                parent.array.key1 = value1, parent.array.key2 = value2
+
+        parent -> array -> [{key1: value1}, {key1: value2}] are flattened as
+                parent.array.key1 = [value1, value2]
+
+        parent -> array -> [ele1, ele2, {key1: value1}] are flattened as
+                'parent.array.0.0': ele1, 'parent.array.0.1': ele2, 'parent.array.key1': value1
+
+        parent -> array -> [[ele1, ele2], [ele3]] are flattened as
+                'parent.array.0.0': ele1, 'parent.array.0.1': ele2, 'parent.array.1.0': ele3
+
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is not dict:
+            raise PostprocessError(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                "Elasticsearch _source value is not a json",
+            )
+
+        def recurse(es_json: JSON, prefix: AnyStr, out: JSON) -> JSON:
+            if type(es_json) is dict:
+                for key in es_json:
+                    recurse(es_json[key], prefix + key + ".", out)
+            elif type(es_json) is list:
+                for i, element in enumerate(es_json):
+                    if type(element) is dict:
+                        recurse(element, prefix, out)
+                    else:
+                        recurse(element, prefix + str(i) + ".", out)
```

Comment:
So, here's the question:  why do we need to add the list element index to the prefix?  Why can't we simply recurse on each element in the list using the current prefix, like we do if the element is a `dict`?

If the resulting key produces a collision, then there is no problem -- we'll keep the multiple entries in a list under that key.  We [discussed this before](https://github.com/distributed-system-analysis/pbench/pull/2381#discussion_r686201449), but let's recap and update....

This change would mean that, if the input were
```
{
  "inlist": [0, 1, 2, 3]
}
```
then the output would be 
```
{
  "inlist": [0, 1, 2, 3]
}
```
(i.e., the same as the input), and I think that that is a Good Thing, that it is better than what the code currently produces, which would be something like
```
{
  "inlist.0": 0,
  "inlist.1": 1,
  "inlist.2": 2,
  "inlist.3": 3
}
```
where the index suffixes don't really add any value.

I think this change does exactly the right thing in the list-of-objects case:
```
{
  "inlist": [ { "inlist0": [0, 1]}, { "inlist0": [0, 1]}, ]
}
```
which should produce
```
{
  "inlist.inlist0": [0, 1, 0, 1]
}
```

However, in the list-of-lists case, things get interesting.  For:
```
{
  "inlist": [ [0, 1], [1, 2], [2, 3] ]
}
```
The code that I'm currently proposing would produce
```
{
  "inlist": [ 0, 1, 1, 2, 2, 3 ]
}
```
If we don't like that, we _could_ insert the list index only if the list element is itself a list, which would produce something like this:
```
{
  "inlist.0": [0, 1],
  "inlist.1": [1, 2],
  "inlist.2": [2, 3]
}
```
That maintains more of the original structure, but I don't think it's a useful improvement over the single, combined list -- once again, the index values in the keys don't really offer much (especially since we cannot count on the ordering), so I'm not inclined toward this direction.

So, I propose that we remove the `if` statement at line 93 and the `else` clause (and the `i` variable and the `enumerate()` call).

Reply:
That's a good point, but I am not sure whether we know the _right_ answer at the moment, I think there are multiple valid answers here but until we explore this little bit more with dashboard in play we wouldn't know. I do like your suggestion and it simplifies the code even more. Sure we can always come back and update the code to what we think is the correct formatting, and if this simplifies the code for now then +1 from me. @dbutenhof anything to add?

Reply:
I have no idea what the current page render will do with a leaf array. On the other hand, we know that using the flattened index-qualified path will look weird, too. Neither really makes sense given the current dashboard implementation, which we know is wrong.

As a result, I would like to very loudly and uncertainly state that I have no opinion, your opinions are almost certainly wrong. and, operationally, I have no idea what any of that means for this PR. 😬 

I think we really need to try to unravel this with Aquib and Anisha, as Nikhil started to plan to try to do at the last dashboard sync. Then we can maybe have some idea of how to move forward.

Reply:
Apparently I "own" this "conversation".  I will mark it resolved when either (a) I see this code replaced with code which returns the raw (or something close to it) ES output, as per Peter's suggestion, or (b) when Nikhil proffers an argument that the current code (possibly as amended) is no worse than any other option at this juncture.

Reply:
I will update the code to return the raw ES output with only specified fields hierarchy.